### ISSUE TITLE
perf(evals): close the ~4.3× evmap gap vs native pil2 (cubic axis-reduction)

### ISSUE THREAD

**Issue by erkhesbold2005-blip** [fractalyze/zisk-zorch#68]
## Goal

Close the **~4.3× gap** on the evals stage (`deep/opening.py::open_columns`, pil2's `evmap`) vs the native pil2-stark CUDA prover.

## Why

Baselining the inner-proof stages leaves evals as a real (if minor-sized) gap: it's the out-of-domain opening — dotting the precomputed LEv vector against each committed column, a batched Fp3-weighted reduction over the base domain. It runs every proof and is pure overhead if slower than native.

## Measured (RTX 5090, M=68 committed columns, 1 opening point)

| | @ N=2²² | per-element |
|---|---|---|
| **pil2 `evmap`** (`computeEvals_v2` + `computeEvalsReduction`) | **3.70 ms** | 0.88 ns |
| **ours** (`open_columns`, extrapolated) | **~16 ms** | ~3.9 ns |
| ratio | **~4.3× slower** | |

Ours measured at 1.93 ms @N=2¹⁹, 3.76 @2²⁰, 8.12 @2²¹ — ~constant 3.9 ns/element (no utilization gain), extrapolating to ~16 ms at the real N=2²². pil2's number is a fresh same-GPU run of the lifted kernel (`/tmp/claude-1006/evmap_bench`).

## Root cause

`open_columns` computes `jnp.sum(lev_per_col * base_coset, axis=0)` — a reduction over the **base-domain axis (N=2²²)** of the cubic committed columns. This is XLA's **cubic axis-reduction**, the same "reduction over the cubic dtype lowers poorly" weakness found in the logup grand-sum (#64). pil2's `computeEvals_v2` instead runs an optimized **shared-memory 16-way tree reduction** (one block per (column, opening-point) eval, `blockIdx.y` splitting the N-row reduction into 16 slices, `Goldilocks3::mul`-accumulate into `__shared__`, tree-reduce) — `starks_gpu.cu:513-599`. That's the 4.3× (0.88 vs 3.9 ns/element).

Unlike logup's fold (small static I=8, closed by unrolling), evals reduces over the **large** N axis, so the fix is a better cubic reduction, not an unroll.

## Levers

1. **Restructure as a matvec.** `sum(lev[:,None] * cols, axis=0)` is `lev @ cols`. If the cubic dtype supports `jnp.einsum`/`dot`, XLA's dot/reduce path may beat the elementwise-mul-then-`jnp.sum`. (Check whether the custom cubic dtype has a dot lowering; if not, this is a no-op.)
2. **Cubic-reduction codegen.** The core issue is XLA's reduction over the 3-limb cubic element; the SoA cubic layout (xla_fork#258) raises the reduction's memory-bandwidth ceiling, and a proper tree/segmented reduction lowering would close most of the gap.
3. A custom reduction primitive in **zorch** (`open_columns` lives there / in the deep glue) mirroring pil2's shared-mem tiled reduction.

## End criteria

- [x] `open_columns` drops from ~16 ms toward pil2's 3.70 ms at N=2²² (target: within ~1.5×). — **3.06 ms / 0.82× at N_ext=2²³ (zorch#512), faster than native**
- [x] `deep/opening_test.py` (evmap byte-match) still passes. — byte-identical (zorch `deep_test` green)

## Repro

- Ours: time `deep/opening.py::open_columns` jitted at n_bits 19–21, M=68, extrapolate (2²²×M cubic OOMs; measure ≤2²¹). Bench: `/tmp/claude-1006/gsumbench/deep_evals_bench.py`.
- pil2: `LD_LIBRARY_PATH=/tmp/claude-1006/gmp-prefix/lib /tmp/claude-1006/evmap_bench` = 3.70 ms (M=68×1 opening, N=2²²); source `/tmp/claude-1006/gsumbench/evmap_bench.cu`, build `/tmp/claude-1006/build_run_evmap.sh`.
- ⚠️ env: needs `CUDA_VISIBLE_DEVICES=0`.

**Comment by erkhesbold2005-blip** [comment]
## Reproduced — and the extrapolation holds, measured directly

Re-run 2026-07-15 on the same RTX 5090 (driver 595.71.05), branch
`fix/deep-coset-compile` (this issue's code, i.e. #61's wire branch + #67's
`f435cba`), against `origin/main`'s pins (`zorch==0.1.0.dev20260714055619`,
`jax==jaxlib==jax-cuda12-pjrt==0.10.0.dev20260714051543`, `zk-dtypes==0.0.10`) —
no local override, no swapped plugin.

**The ~4.3× is confirmed, and it no longer needs to be an extrapolation:**

| | this issue | measured |
|---|---|---|
| native `evmap` (1 opening, N=2^22) | 3.70 ms | **3.72 ms** |
| ours `open_columns` @ n_bits=22 | ~16 ms (extrapolated) | **15.80 ms** (measured) |
| ratio | ~4.3× | **4.25×** |

Sub-target points reproduce too: n_bits 19 / 20 / 21 → **1.90 / 3.73 / 8.01 ms**
(this issue recorded 1.93 / 3.76 / 8.12).

### The OOM that forced the extrapolation is removable

`2^22 × M cubic OOMs` is not the card. JAX caps the BFC allocator at ~75% of the
GPU — **23.5 GiB of this 5090's 31.8 GiB**. The working set at n_bits=22 is
~25.5 GiB (the 12.75 GiB committed buffer plus one more of the same size), which
lands in the band between the cap and the hardware. With
`XLA_PYTHON_CLIENT_MEM_FRACTION=0.95` it simply runs:

```sh
CUDA_VISIBLE_DEVICES=0 XLA_PYTHON_CLIENT_PREALLOCATE=false \
XLA_PYTHON_CLIENT_MEM_FRACTION=0.95 PYTHONPATH=<zz-deep> \
  python /tmp/claude-1006/gsumbench/deep_evals_bench.py 22
```

Worth using for any future number here — extrapolation is what made #66's
quotient figure wrong by ~3.5×, so measuring at the real size is cheap insurance.

### Two footguns for whoever picks this up

- **`evmap_bench` prints two cases** — 1 opening point, then 2 (4.98 ms). This
  issue's 3.70 ms is the **1-opening** block; reading the last block instead is a
  ~1.34× scope error.
- **This issue indexes by `n_bits`; #69 indexes by `n_ext`.** The "2^19/2^20/2^21"
  here are n_bits (= n_ext 2^20/2^21/2^22). `deep_evals_bench.py`'s argument is
  n_bits and its printout is n_ext.

### On the gap itself

Our committed buffer is 68 **cubic** columns = 13.69 GB, where the native holds
62 base + 6 cubic = 80 gl64 = 5.37 GB — a **2.55× read amplification** before any
kernel runs. That accounts for part of the 4.25×; the rest is consistent with the
cubic axis-reduction this issue names.

Recorded in `docs/zisk-baseline.md` (#60).

**Comment by erkhesbold2005-blip** [comment]
### Session 2026-07-22 progress — lever 3 built, measured 4.25x → 1.12x

- Done: root cause isolated and fixed as a cross-repo pair, both PRs open:
  - **The gap is the accumulator dtype, not the loads.** A weighted gl64 column reduce runs at 1.36 TB/s on this card; the same reduce with a cubic accumulator runs at 154 GB/s (9x). A field sum over the domain is limb-wise, so each opening's 3 output limbs are 3 *base* weighted sums — sibling reduces over one block, which XLA multi-output fusion merges into a single pass holding 3 scalar accumulators (pil2's exact register structure — this is lever 3 without a hand-written kernel).
  - **The only thing blocking that fusion was an XLA estimate bug**: the shared-memory estimate for column reduces assumes vector size 4 for every dtype (33 KB/reduce), while the emitter loads 64-bit types unvectorized (8.4 KB actual). Two siblings "exceeded" the 48 KB budget for kernels that need 25 KB. Fixed in **fractalyze/xla#306** (estimate keyed on the fusion's smallest input width + a merge-stability guard for mixed-width siblings; caught a real 49,920 B ptxas overflow).
  - **`open_columns` rewritten in the limb-decomposed sibling form** in **fractalyze/zorch#512** (on top of #496's `pcs/deep.py`): degree-generic, grouped by opening runs, phase-widened ext block; byte-identical by exactness of field addition, pinned by a new test vs the naive per-column dot.
- Measured (RTX 5090, 62 base + 6 cubic, stride 2, N=2^22, 1 opening; direct, no extrapolation):
  | | ms | vs pil2 3.70 |
  |---|---|---|
  | issue baseline (all-cubic, as filed) | 15.80 | 4.25x |
  | #496 base/ext split alone | 7.79 | 2.11x |
  | zorch#512 + xla#306 | **4.14** | **1.12x** |
  2-pt mixed case: 4.99 ms. On today's wheel (no xla fix) zorch#512 is neutral (7.6 ms), so it lands independently. zisk goldens all green under the patched plugin (trace-commit root, linear hash, LEv, FRI fold, poseidon2).
- Next session: land chain is xla#306 → plugin wheel, zorch#512 → zorch wheel, then the zisk deep-via-zorch swap picks both up; then re-measure here and tick the end criteria. Remaining headroom (optional): phases sweep on the ext block (~-0.3 ms), single-pass multi-opening for the 2-pt case (~-15%), and the systemic cubic-SoA lever stays with fractalyze/xla_fork#258.
- Blocker: none.

**Comment by erkhesbold2005-blip** [comment]
### Resolved — evals now beats native pil2, and the whole compiler line it spawned is retired

Final state: `open_columns` written as **batched block reductions** (fractalyze/zorch#512) lands *below* pil2 on both time and memory, at N_ext=2^23 on the RTX 5090, byte-identical (pinned against the plain per-column dot + the pil2 golden).

**Time**

| shape | zorch#512 | native pil2 `evmap` | ratio |
|---|---|---|---|
| N_ext=2^23, 1 opening | **3.06 ms** | 3.74 ms | **0.82×** |
| N_ext=2^23, 2 openings | **3.06 ms** | 5.07 ms | **0.60×** |

Our stage is opening-count-invariant (always two block reductions, base + ext); pil2 re-runs `computeEvals` per opening, so we pull further ahead as opening points grow.

**Memory** — neither materializes an (N, M) intermediate; both are bound by the committed columns (resident anyway). Ours is marginally smaller (5.00 vs 5.37 GiB @ 2^23) because base columns stay base (#69 split) instead of pil2's uniform 80-gl64 buffer. No OOM at 2^24 (15.5 GiB peak).

**The result vs where this issue opened:** 15.8 ms / 4.25× → **3.06 ms / 0.82×** — from ~4× slower to ~1.2× faster than native.

**Why the simple form wins, and why the fancy fixes were retired.** The premise of this issue — "cubic axis-reduction lowers ~9x poorly" (my 154 GB/s floor probe) — was measured on an early-July frx build. Current frx lowers a cubic block reduction near bandwidth (~1223 GB/s), so the cliff is **fixed upstream**. That obsoleted the whole optimization line built against it:
- The per-limb decomposition (this PR's first approach) is now *slower* than the plain block reduce — dropped.
- fractalyze/xla#315 (extension-reduce decomposer): closed won't-do — measured net-negative; the cliff it targeted is gone.
- fractalyze/xla#306 (sibling-fusion shmem-estimate fix): a correct general fix, kept, but no longer needed for this stage.

The per-column slowness was never the cubic accumulator — it was 68 separate reduce launches. Writing it as two block reductions (a static per-column weight gather + `sum(axis=0)`) collapses 68 kernels → 4, which is the entire win. ~10 lines, no compiler change.

**End criteria** — perf target met (0.82× ≤ 1.5×); byte-match holds (`deep_test` green; zisk `opening_test` byte-matches once it consumes the new zorch). 

**Landing gate (all that remains):** zorch#512 merges → new zorch wheel → bump zisk's zorch pin; zisk's `deep/fri_polynomial.open_columns` already calls through to `zorch.pcs.deep.open_columns`, so no zisk code change beyond the pin.

**Comment by erkhesbold2005-blip** [comment]
### Session 2026-07-27 progress — landing chain has one more link (a #512 bug surfaced during the bump)

- Done:
  - zorch#512 (block-form `open_columns`) **merged** (`dd37b25`), wheel `pyzorch-0.1.0.dev20260727025554` published.
  - Bumping the consumer (zisk#97) to that wheel surfaced a regression **in the merged #512**: `open_columns` gathers per-column weights with `weights[:, np.array(opening_pos[:b])]`; when a block is empty (wired DEEP opens a lone cubic quotient column, `base_cols` is `(N,0)`), `np.array([])` is float64 → frx rejects it as an index. #512's tests always had both a base and an ext block, so it slipped.
  - Fixed forward in **zorch#539** (`3351c58`): pin the gather to `np.intp` + a single-field-block regression test (all-base and all-ext). 5/5 zorch deep tests green, black-clean, pushed. **USER merges.**
  - zisk#97 (the #519 claim-reduction bump) is **unaffected** and stays green on `dev20260726114757` — verified `prover_test` / `fri_polynomial_test` / `lev_test` pass against both wheels; only `opening_test` trips the empty-block bug that #539 resolves. Did NOT push a broken bump to #97; posted the status on #97 instead.
- Next session: after zorch#539 merges → cut a pyzorch wheel carrying it → one-line pin bump on zisk#97 (`requirements.in` + `bazel run //:requirements.update`) to that wheel; then `bazel test //zisk_zorch/deep/... //zisk_zorch:prover_test` must be green (esp. `opening_test`). No zisk code change — `deep/fri_polynomial.open_columns` already calls through to `zorch.pcs.deep`. Then this issue's perf lands end-to-end (3.06 ms / 0.83× pil2) and #68 can close.
- Blocker: zorch#539 merge + a wheel carrying it (both USER/CI-gated).
